File: LeetCode/python/938.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rangeSumBST(self, root, L, R):
        """
        :type root: TreeNode
        :type L: int
        :type R: int
        :rtype: int
        """
        if not root:
            return 0
        sum = 0
        if root.val <= R and root.val >= L:
            sum += root.val
        if root.val < R:
            sum += self.rangeSumBST(root.right, L, R)
        if root.val > L:
            sum += self.rangeSumBST(root.left, L, R)
        return sum

File: LeetCode/cpp/938.cpp
class Solution {
    int sum;
public:
    int rangeSumBST(TreeNode *root, int &L, int &R) {
        sum = 0;
        DFS(root, L, R);
        return sum;
    }

    void DFS(TreeNode *node, int &L, int &R) {
        if (!node)
            return;
        if (node->val >= L && node->val <= R)
            sum += node->val;
        if (node->val < R)
            DFS(node->right, L, R);
        if (node->val > L)
            DFS(node->left, L, R);
    }
};
